editor/src/components/inspector/common/name-and-icon-hook.ts

```diff
@@ -30,20 +32,33 @@ export interface NameAndIconResult {
 }
 
 export function useNameAndIcon(path: TemplatePath): NameAndIconResult {
-  const metadata = useEditorState(
-    (store) => store.editor.jsxMetadataKILLME,
-    'useNameAndIcon metadata',
-  )
-  const components = useEditorState(
-    (store) => getOpenUtopiaJSXComponentsFromState(store.editor),
-    'useNameAndIcon components',
-  )
-  const imports = useEditorState(
-    (store) => getOpenImportsFromState(store.editor),
-    'useNameAndIcon imports',
+  return useEditorState(
+    (store) => {
+      const metadata = store.editor.jsxMetadataKILLME
+      const components = getOpenUtopiaJSXComponentsFromState(store.editor)
+      const imports = getOpenImportsFromState(store.editor)
+      return getNameAndIconResult(path, components, metadata, imports)
+    },
+    'useNameAndIcon',
+    (oldResult, newResult) => {
+      if (!TP.pathsEqual(oldResult.path, newResult.path)) {
+        return false
+      }
+      if (oldResult.label !== newResult.label) {
+        return false
+      }
+      if (!shallowEqual(oldResult.iconProps, newResult.iconProps)) {
+        return false
+      }
+      if (
+        oldResult.name?.baseVariable !== newResult.name?.baseVariable ||
+        !PP.pathsEqual(oldResult.name?.propertyPath ?? null, newResult.name?.propertyPath ?? null)
```

Comment:
`!..?.??,.?.??`

Reply:
omg please no! Trying to parse this is just making my brain give up. We had a fairly similar situation in the unsaved files work that lead to us pulling apart the conditions [here](https://github.com/concrete-utopia/utopia/pull/950/files#diff-9b4470f22a98015a2beb08ed997473a32b34aacd8a3bf1a7e748f7018e507ab6R194-R208) ([here is a diff with the previous version](https://github.com/concrete-utopia/utopia/pull/950/commits/34bf8bedfeb7b9403805b7ebef4508e900843bd7))